### ISSUE TITLE
live grep gets stuck when entering first several characters

### ISSUE THREAD

**Issue by bryceyang** [nvim-telescope/telescope.nvim#269]
### Description

live grep gets stuck when entering first serval characters.
![image](https://user-images.githubusercontent.com/9302617/99867132-adec7680-2bf1-11eb-99f2-e1661e6f158e.png)

But using fzf with rg is very smooth.

**Expected Behavior**

live grep should be fast and no noticable lag when input.

**Actual Behavior**

input could stuck when using live grep

### Details

<details><summary>Reproduce</summary>

1. nvim -NU test.vim
2. :Telescope live_grep
3. input some charaters, problems appears
</details>

<details><summary>Environment</summary>

- nvim version: 
```shell
NVIM v0.5.0-828-g0a95549d6
Build type: Release
LuaJIT 2.1.0-beta3
Compilation: /Applications/Xcode-10.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/clang -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -DNDEBUG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/Users/travis/build/neovim/bot-ci/build/neovim/build/config -I/Users/travis/build/neovim/bot-ci/build/neovim/src -I/Users/travis/build/neovim/bot-ci/build/neovim/.deps/usr/include -I/usr/include -I/usr/local/opt/gettext/include -I/Users/travis/build/neovim/bot-ci/build/neovim/build/src/nvim/auto -I/Users/travis/build/neovim/bot-ci/build/neovim/build/include
Compiled by travis@Traviss-Mac-6.local

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/share/nvim"

Run :checkhealth for more info
```
- Operating system: 
mac os big sur 
![image](https://user-images.githubusercontent.com/9302617/99867160-dd02e800-2bf1-11eb-8d5d-522f85854d38.png)

```shell
Darwin bogon 20.1.0 Darwin Kernel Version 20.1.0: Sat Oct 31 00:07:11 PDT 2020; root:xnu-7195.50.7~2/RELEASE_X86_64 x86_64
```
- Telescope commit: 
`929a303`
</details>

<details><summary>Configuration</summary>
<p>
 

```viml
set nocompatible hidden laststatus=2

if !filereadable('/tmp/plug.vim')
  silent !curl --insecure -fLo /tmp/plug.vim
    \ https://raw.githubusercontent.com/junegunn/vim-plug/master/plug.vim
endif

source /tmp/plug.vim
call plug#begin('/tmp/plugged')
Plug 'nvim-lua/popup.nvim'
Plug 'nvim-lua/plenary.nvim'
Plug 'nvim-telescope/telescope.nvim'
call plug#end()

autocmd VimEnter * PlugClean! | PlugUpdate --sync | close
lua << EOF

require('telescope').setup{
  defaults = {
    vimgrep_arguments = {
      'rg',
      '--color=never',
      '--no-heading',
      '--with-filename',
      '--line-number',
      '--column',
      '--smart-case'
    },
    prompt_position = "bottom", 
    prompt_prefix = ">",
    selection_strategy = "reset",
    sorting_strategy = "descending",
    layout_strategy = "horizontal",
    layout_defaults = {},
    file_ignore_patterns = {},
    shorten_path = true,
    winblend = 0,
    width = 0.75,
    preview_cutoff = 120,
    results_height = 1,
    results_width = 0.8,
    border = {},
    borderchars = { '─', '│', '─', '│', '╭', '╮', '╯', '╰'},
    color_devicons = true,
    use_less = true,
  }
}
EOF
```

**Comment by tami5** [comment]
I can reproduce (and as long as I remember) this lag noticeable, thanks @bryceyang  
@tjdevries when you have time this very important, it also answer the previously  question you've asked :smile: about "missing functionalities" :P .

**Comment by bryceyang** [comment]
ping.. any progress?

**Comment by iverberk** [comment]
Maybe as a first step we could make it configurable when the grepping starts? Or is the problem not in the amount of results returned? I think the delay is due to a lot of results based on 1 or 2 letter searches. It becomes a lot smoother with more letters.

**Comment by tjdevries** [comment]
You can use this: https://github.com/nvim-telescope/telescope-fzf-writer.nvim#exports And it has the configuration you're looking for. I think it's pretty good solution for now.

**Comment by iverberk** [comment]
That works nicely, thanks @tjdevries

**Comment by vpeopleonatank** [comment]
Same issue when open Telescope find_files in folder with multiple large json files. I 've to use vim with vim-clap in this case.  It's smooth and non-blocking when move between large json files with previewing

**Comment by Conni2461** [comment]
That happens because previewer highlighting still happen in main thread (file reading is async). In https://github.com/nvim-telescope/telescope.nvim/commit/1d40ab5ccda6be3680fd809d34d93366b11c4ef8 i've added a way to disable this manually. 

Example: This will disable highlighting for `.lua` and `.json`
```lua
local previewers = require('telescope.previewers')

local _bad = { '.*%.json', '.*%.lua' } -- Put all filetypes that slow you down in this array
local bad_files = function(filepath)
  for _, v in ipairs(_bad) do
    if filepath:match(v) then
      return false
    end
  end

  return true
end

local new_maker = function(filepath, bufnr, opts)
  opts = opts or {}
  if opts.use_ft_detect == nil then opts.use_ft_detect = true end
  opts.use_ft_detect = opts.use_ft_detect == false and false or bad_files(filepath)
  previewers.buffer_previewer_maker(filepath, bufnr, opts)
end

require('telescope').setup {
  defaults = {
    buffer_previewer_maker = new_maker,
  }
}
```

I will add some of these configurations to the wiki at some point :)

The advantage of this solution is that you can also exclude files based on size. So if files > 1mb make problems, get file size and don't highlight, or don't read them at all.

**Comment by Conni2461** [comment]
Tj made some progress regarding live_grep. See https://github.com/nvim-telescope/telescope.nvim/pull/987

**Comment by Conni2461** [comment]
This issue was fixed in #987 

Closing for now. If you still can reproduce it feel free to open a new issue. Thanks :)
